platform/stm32f4xx/export/platform/i2c_bus.hpp

```diff
@@ -0,0 +1,648 @@
+    constexpr auto init_const_obj = i2c_config::m_init_obj;
+    auto init_obj = init_const_obj;
+    I2C_Init(i2c, &init_obj);
+
+    I2C_Cmd(i2c, ENABLE);
+
+    if (i2c_config::m_mode == i2c_mode::IRQ) {
+        constexpr auto irqn_ev = pick_ev_irqn();
+        constexpr auto irqn_er = pick_er_irqn();
+
+        auto lambda_ev = [this]() {
+            this->irq_ev_handler();
+        };
+
+        auto lambda_er = [this]() {
+            this->irq_er_handler();
+        };
+
+        IRQ_manager::subscribe(irqn_ev, lambda_ev);
+        IRQ_manager::subscribe(irqn_er, lambda_er);
+    }
+
+    set_inited();
+
+    return ecl::err::ok;
+}
+
+template< class i2c_config >
+void i2c_bus< i2c_config >::set_rx(uint8_t *rx, size_t size)
+{
+    ecl_assert(inited());
+
+    m_rx = rx;
+    m_rx_size = size;
+    m_rx_left = m_rx_size;
+}
+
+template< class i2c_config >
+void i2c_bus< i2c_config >::set_tx(const uint8_t *tx, size_t size)
+{
+    ecl_assert(inited());
+
+    m_tx = tx;
+    m_tx_size = size;
+    m_tx_left = m_tx_size;
+}
+
+template< class i2c_config >
+void i2c_bus< i2c_config >::set_tx(size_t size, uint8_t fill_byte)
+{
+    ecl_assert(inited());
+
+    (void) size;
+    (void) fill_byte;
+}
+
+template< class i2c_config >
+ecl::err i2c_bus< i2c_config >::do_xfer()
+{
+    ecl_assert(inited());
+
+    if (i2c_config::m_mode == i2c_mode::POLL) {
+        if (m_tx) {
+            i2c_transmit_poll();
+            m_event_handler(channel::tx, event::tc, 0);
+        }
+
+        if (m_rx) {
+            i2c_receive_poll();
+            m_event_handler(channel::rx, event::tc, 0);
+        }
+
+        m_event_handler(channel::meta, event::tc, 0);
+    } else if (i2c_config::m_mode == i2c_mode::IRQ) {
+        // There are common interrupts for tx and rx.
+        // If no tx is required then rx will be done,
+        // otherwise rx will be set in ISR, when tx
+        // will be finished
+        if (m_tx) {
+            m_direction = MASTER_TX;
+        } else if (m_rx) {
+            m_direction = MASTER_RX;
+        }
+
+        constexpr auto irqn = pick_ev_irqn();
+        i2c_setup_xfer_irq();
+        IRQ_manager::unmask(irqn);
+    }
+
+    return ecl::err::ok;
+}
+
+template< class i2c_config >
+void i2c_bus< i2c_config >::reset_buffers()
+{
+    m_tx = nullptr;
+    m_rx = nullptr;
+    m_tx_left = m_tx_size = 0;
+    m_rx_left = m_rx_size = 0;
+}
+
+template< class i2c_config >
+void i2c_bus< i2c_config >::set_handler(const handler_fn &handler)
+{
+    m_event_handler = handler;
+}
+
+template< class i2c_config >
+ecl::err i2c_bus <i2c_config>::i2c_setup_xfer_irq()
+{
+    constexpr auto i2c = pick_i2c();
+
+    while (I2C_GetFlagStatus(i2c, I2C_FLAG_BUSY) == SET);
+    I2C_GenerateSTART(i2c, ENABLE);
+
+    // TODO Add error handling and enable error irq
+    I2C_ITConfig(i2c, I2C_IT_EVT | I2C_IT_BUF, ENABLE);
+
+    return ecl::err::ok;
+}
+
+template< class i2c_config >
+ecl::err i2c_bus <i2c_config>::i2c_transmit_poll()
```

Comment:
agree, but it will make sense after adding error handling. For polling it just return on timeout I believe.